### PR TITLE
Fixed #20892 -- Allowed configuring memcached client using OPTIONS

### PR DESCRIPTION
**Pull request by edmorley**
Previously the `django.core.cache.backends.memcached.MemcachedCache` backend would ignore the contents of `OPTIONS`, and the `django.core.cache.backends.memcached.PyLibMCCache` backend would use them to set pylibmc behaviors instead.

Now, the memcached backends pass the contents of `OPTIONS` as keyword arguments to the client constructors, allowing for more advanced control of client behavior.

Passing pylibmc behavior settings as top-level attributes of `OPTIONS` is still supported until Django 2.1, but will generate a deprecation warning. They should now be set under a `behaviors` key within `OPTIONS` instead.

Tested against both the MemcachedCache and PyLibMCCache backends on Python 2.7 using:
`./runtests.py -v 2 cache.tests.MemcachedCacheTests` 

Fixes:
https://code.djangoproject.com/ticket/20892
https://code.djangoproject.com/ticket/15815

Supersedes PR #6233.

### CODE REVIEW

File: docs/topics/cache.txt [hunk 1]
```diff
@@ -401,6 +403,19 @@ behavior. These arguments are provided as additional keys in the
     On some backends (``database`` in particular) this makes culling *much*
     faster at the expense of more cache misses.
 
+  .. versionchanged:: 1.11
```
Comment (timgraham) [inline docs/topics/cache.txt]: This note should be self-contained as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features

File: docs/topics/cache.txt [hunk 2]
```diff
@@ -435,8 +450,39 @@ of 60 seconds, and a maximum capacity of 1000 items::
         }
     }
 
-Invalid arguments are silently ignored, as are invalid values of known
-arguments.
+Here, a python-memcached based backend is being configured with an object size
```
Comment (timgraham) [inline docs/topics/cache.txt]: Use double backticks around python-memcached and pylibmc to fix the spelling issue and avoid having to add them to the wordlist.

**Comment by timgraham** [comment]
Jenkins currently uses this configuration:
```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.locmem.LocMemCache',
    },
    'memcached': {
        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
        'LOCATION': '127.0.0.1:%s' % memcached_port,
        'KEY_PREFIX': '%s_%s_%s' % (PY_VERSION, DATABASE, ASCII_BUILD_NAME),
    }
}
```
It would be nice to have pylibmc in there also so the tests run with both backends. Do you think it's feasible? If so, we can create a separate ticket, perhaps to precede this PR.

File: docs/topics/cache.txt (hunk 1 above)
Comment (edmorley) [inline docs/topics/cache.txt]: Ah sorry I missed that part of the guide. So like this?
https://emorley.pastebin.mozilla.org/8905150

**Comment by edmorley** [comment]
Adding Jenkins testing of the pylibmc backend definitely sounds like a sensible medium/long term plan.

It would require a refactor of the memcached tests however, since they currently just pick the first memcached-like backend found, and use that for everything.

File: docs/topics/cache.txt (hunk 1 above)
Comment (timgraham) [inline docs/topics/cache.txt]: Yes

**Comment by edmorley** [comment]
I've opened PR #7168 to allow for the testing of the pylibmc backend on Jenkins.

Let's wait until that is merged before proceeding here, since (a) there will be conflicts, (b) we'll actually be able to check that this PR is working without having to test locally.

File: docs/ref/settings.txt [hunk 3]
```diff
@@ -220,9 +220,14 @@ Extra parameters to pass to the cache backend. Available parameters
 vary depending on your cache backend.
 
 Some information on available parameters can be found in the
-:doc:`Cache Backends </topics/cache>` documentation. For more information,
+:ref:`cache arguments <cache_arguments>` documentation. For more information,
```
Comment (timgraham) [inline docs/ref/settings.txt]: Please make this change a separate commit

File: docs/ref/settings.txt [hunk 4]
```diff
@@ -220,9 +220,14 @@ Extra parameters to pass to the cache backend. Available parameters
...
 consult your backend module's own documentation.
 
+.. versionchanged:: 1.11
```
Comment (timgraham) [inline docs/ref/settings.txt]: This annotation seems out of place since there's no discussion of it otherwise.

File: docs/releases/1.11.txt [hunk 5]
```diff
@@ -166,7 +166,9 @@ Minor features
 Cache
 ~~~~~
 
-* ...
+* Memcached backends now pass the contents of :setting:`OPTIONS <CACHES-OPTIONS>`
+  as keyword arguments to the client constructors, allowing for more advanced
+  control of client behavior.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: An example could be nice

File: docs/topics/cache.txt [hunk 6]
```diff
@@ -401,6 +403,21 @@ behavior. These arguments are provided as additional keys in the
     On some backends (``database`` in particular) this makes culling *much*
     faster at the expense of more cache misses.
 
+  Memcached backends pass the contents of :setting:`OPTIONS <CACHES-OPTIONS>`
+  as keyword arguments to the client constructors, allowing for more advanced
+  control of client behavior. For example usage, see below.
+
+  .. versionchanged:: 1.11
+
+      Memcached backends can now be configured using ``OPTIONS``.
```
Comment (timgraham) [inline docs/topics/cache.txt]: This is a little confusing because right after it you say how configuring pyblic used 'behaviors' in OPTIONS. Add something like "In older versions, you could pass pyblic behavior settings.." (Maybe combine it with the deprecation note)
Also move, these annotations to the bottom of the section.

File: docs/topics/cache.txt [hunk 7]
```diff
@@ -435,8 +452,39 @@ of 60 seconds, and a maximum capacity of 1000 items::
         }
     }
 
-Invalid arguments are silently ignored, as are invalid values of known
-arguments.
+Here, a ``python-memcached`` based backend is being configured with an object
```
Comment (timgraham) [inline docs/topics/cache.txt]: Here's an example configuration for a ``python-memcached`` based backend with an object...

File: docs/topics/cache.txt [hunk 8]
```diff
@@ -435,8 +452,39 @@ of 60 seconds, and a maximum capacity of 1000 items::
...
+        'default': {
+            'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
+            'LOCATION': '127.0.0.1:11211',
+            'OPTIONS': {
+                'server_max_value_length': 1024 * 1024 * 2,
+            }
+        }
+    }
+
+And here, a ``pylibmc`` based backend is being configured to use the binary
```
Comment (timgraham) [inline docs/topics/cache.txt]: similar wording as above

File: docs/topics/cache.txt [hunk 9]
```diff
@@ -435,8 +452,39 @@ of 60 seconds, and a maximum capacity of 1000 items::
...
+                'username': 'user',
+                'password': 'pass',
+                'behaviors': {
+                    'ketama': True,
+                }
+            }
+        }
+    }
+
+For some backends invalid arguments are silently ignored, as are invalid values
```
Comment (timgraham) [inline docs/topics/cache.txt]: backends,

File: tests/cache/tests.py [hunk 10]
```diff
@@ -1259,6 +1268,36 @@ class PyLibMCCacheTests(BaseMemcachedTests, TestCase):
     def test_invalid_key_characters(self):
         pass
 
+    @override_settings(CACHES=caches_setting_for_tests(
+        base=PyLibMCCache_params,
+        exclude=memcached_excluded_caches,
+        OPTIONS={
+            'binary': True,
+            'behaviors': {'tcp_nodelay': True}
```
Comment (timgraham) [inline tests/cache/tests.py]: add trailing comma

File: docs/releases/1.11.txt (hunk 5 above)
Comment (timgraham) [inline docs/releases/1.11.txt]: Or a link to the new examples

**Comment by edmorley** [comment]
Thank you for the feedback - updated :-)

**Comment by timgraham** [comment]
merged in 1d54fb4483f034d2dced86f1b012671c8ee6ef5d and 65ec8fa8ca56a5378345375e1079025c96d0b833, thanks!
